core/trino-main/src/main/java/io/trino/sql/planner/DomainTranslator.java

```diff
@@ -911,8 +911,17 @@ yield or(new ComparisonExpression(EQUAL, symbolExpression, coercedLiteral),
 
         private Optional<Object> floorValue(Type fromType, Type toType, Object value)
         {
-            return getSaturatedFloorCastOperator(fromType, toType)
-                    .map(operator -> functionInvoker.invoke(operator, session.toConnectorSession(), value));
+            Optional<ResolvedFunction> operator = getSaturatedFloorCastOperator(fromType, toType);
+            if (operator.isPresent()) {
+                try {
+                    return Optional.of(functionInvoker.invoke(operator.get(), session.toConnectorSession(), value));
+                }
+                catch (TrinoException ignored) {
+                    // Can fail in case of an invalid cast.
+                    // In this case, it's safe to just bail out and leave the cast to be handled later.
```

Comment:
add a test in TestDomainTranslator

Reply:
Added a new test `testFromCoercionComparisonsWithNaN` under  the existing `testFromBasicComparisonsWithNaN`